In [8]:
import pandas as pd
import matplotlib.pyplot as plt

In [9]:
shooters = pd.read_csv('Data/Stanford_MSA_Database.csv')

OK, so there's lots of things we can do with the dataset given that it's a full 335 rows with 55 columns of descriptive text data each. Let's start by coming up with some basic questions we want answered:
- What kinds of weapons are most popular, and where?
- Where are guns most readily accessible in the country?
- How do warning signs change with age and other factors?
- Does lethality change with age?
- Have shootings become, on the whole, more lethal over time?
- How do the number of fatalities compare to the total number of victims?
- Has the number of guns used in a shooting changed on average over time?
- What happens to the shooter, and how does that change with age and race? Does their fate change substantially in incidences of school shootings?
- How do the motivations of mass shooters differ with geography?
- Does military experience alter the motive?

In [10]:
shooters

CaseID                                              Title  \
0         1                      University of Texas at Austin   
1         2                         Rose-Mar College of Beauty   
2         3                       New Orleans Police Shootings   
3         4                     Clara Barton Elementary School   
4         5                                  Olean High School   
5         6               Los Angeles Computer Learning Center   
6         7                                Cal State Fullerton   
7         8                 Grover Cleveland Elementary School   
8         9                       University of South Carolina   
9        10                                 Valley High School   
10       11                              Welding shop in Miami   
11       12                            Wah Mee Club in Seattle   
12       13                               Johnston Post Office   
13       14                      49th Street Elementary School   
14       15                             Ianni's Club in Dallas   
15       16                McDonald's restaurant in San Ysidro   
16       17                         Goddard Junior High School   
17       18                                Atlanta Post Office   
18       19                         New York Technical College   
19       20                    Post office in Edmond, Oklahoma   
20       21                          Fergus County High School   
21       22                          Palm Bay Shopping Centers   
22       23                          Pinellas Park High School   
23       24           Electromagnetic Systems Laboratory (ESL)   
24       25                    Hubbard Woods Elementary School   
25       26             Comet Auto Parts and Montefiore School   
26       27                          Oakland Elementary School   
27       28                   New Orleans Downtown Post Office   
28       29                        Cleveland Elementary School   
29       30                           Orange Glenn Post Office   
..      ...                                                ...   
305     315                     Forestville, Maryland Drive-by   
306     316                              Ferguson, MO Drive by   
307     317                          Appling, GA family murder   
308     318                           Auburn, AL club shootout   
309     319              Nashville Music City Central shooting   
310     320                    Murfreesboro, TN party shooting   
311     321              Florida Ex-wife, ex-girlfriend murder   
312     322                Katy, Texas work retaliation murder   
313     323          Card game ambush in Shreveport, Louisiana   
314     324                Allapattah, Florida street shooting   
315     325  Federal Security Officer has day shooting ramp...   
316     326                Montgomery, Alabama street shooting   
317     327                            Detroit street shooting   
318     328                           Austin Taco truck murder   
319     329                 Birmingham Domestic Dispute Murder   
320     330                            Chicago bus stop ambush   
321     331                      Chicago Family Murder Suicide   
322     332         Phoenix girlfriend murder and cop shootout   
323     333     Friend murder and fire coverup in Moultrie, GA   
324     334     Trailer park triple homocide in South Carolina   
325     335                          Graduation party drive by   
326     336                                Stockton, CA murder   
327     337                               NYC Concert shooting   
328     338                        Las Vegas sidewalk shooting   
329     339                       Army vet shooting in Houston   
330     340                                Fort Warne drive by   
331     341                          Cape Coral Shooting Spree   
332     342                         Webster, MN party shooting   
333     343                         Orlando N

In [11]:
len(shooters)

335

In [12]:
list(shooters)

['CaseID',
 'Title',
 'Location',
 'City',
 'State',
 'Latitude',
 'Longitude',
 'Number of Civilian Fatalities',
 'Number of Civilian Injured',
 'Number of Enforcement Fatalities',
 'Number of Enforcement Injured',
 'Total Number of Fatalities',
 'Total Number of Victims',
 'Description',
 'Date',
 'Day of Week',
 'Date - Detailed',
 'Shooter Name',
 'Number of shooters',
 'Shooter Age(s)',
 'Average Shooter Age',
 'Shooter Sex',
 'Shooter Race',
 'Type of Gun - Detailed',
 'Type of Gun - General',
 'Number of Shotguns',
 'Number of Rifles',
 'Number of Handguns',
 'Total Number of Guns',
 'Number of Automatic Guns',
 'Number of Semi-Automatic Guns',
 'Fate of Shooter at the scene',
 'Fate of Shooter',
 "Shooter's Cause of Death",
 'School Related',
 'Place Type',
 'Relationship to Incident Location',
 'Targeted Victim/s - Detailed',
 'Targeted Victim/s - General',
 'Possible Motive - Detailed',
 'Possible Motive - General',
 'History of Mental Illness - Detailed',
 'History of Mental

I'm curious what's going to happen if I run a market basket analysis. Let's clean up the data a little more and see what happens. I'm working off of the following tutorial:
http://pbpython.com/market-basket-analysis.html

In [13]:
cleaned_shooters = shooters.copy(deep=True)

#Clean gun types
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.strip()
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.lower()
cleaned_shooters['Type of Gun - General'].unique()

array(['multiple guns', 'handgun', 'shotgun', 'rifle', 'unknown',
       'semi-automatic rifle', '9-mm'], dtype=object)

In [14]:
#Clean gun counts
guncount_types = ['Number of Shotguns','Number of Handguns','Number of Rifles']
for guncount_type in guncount_types:
    cleaned_shooters[guncount_type].replace('Unknown', 0, inplace=True)

In [15]:
#Clean military experience
cleaned_shooters['Military Experience'] = cleaned_shooters['Military Experience'].str.lower()
cleaned_shooters['Military Experience'].replace('yes', True, inplace=True)
cleaned_shooters['Military Experience'].replace(['no', 'unknown'], False, inplace=True)

In [16]:
#Set civilian + law enforcement casualties, injuries to bools
cleaned_shooters['Number of Civilian Fatalities'] = cleaned_shooters['Number of Civilian Fatalities'] > 0
cleaned_shooters['Number of Civilian Injured'] = cleaned_shooters['Number of Civilian Injured'] > 0
cleaned_shooters['Number of Enforcement Fatalities'] = cleaned_shooters['Number of Enforcement Fatalities'] > 0
cleaned_shooters['Number of Enforcement Injured'] = cleaned_shooters['Number of Enforcement Injured'] > 0
cleaned_shooters['High Civilian Casualties'] = (shooters['Number of Civilian Fatalities'] + shooters['Number of Civilian Injured']) > 10

In [17]:
#Corner cases (from errors)
cleaned_shooters['Number of Shotguns'].replace('Handgun', 0, inplace=True)
cleaned_shooters['Number of Rifles'].replace('0 (1)', 0, inplace=True)
cleaned_shooters['Number of Handguns'].replace('2 (1)', 0, inplace=True)

In [18]:
#Set gun types to bools
cleaned_shooters['Number of Shotguns'] = pd.to_numeric(cleaned_shooters['Number of Shotguns']) > 0
cleaned_shooters['Number of Rifles'] = pd.to_numeric(cleaned_shooters['Number of Rifles']) > 0
cleaned_shooters['Number of Handguns'] = pd.to_numeric(cleaned_shooters['Number of Handguns']) > 0

In [19]:
#Clean fate of shooter results
cleaned_shooters['Fate of Shooter'].replace('Custody / Escaped', 'Custody/Escaped', inplace=True)

In [20]:
len(cleaned_shooters.loc[
    (cleaned_shooters['Fate of Shooter'].isnull()) | 
    (cleaned_shooters['Fate of Shooter'] == 'FALSE')
])

3

In [21]:
#Drop the three where we don't know the fate of the shooter
cleaned_shooters = cleaned_shooters.loc[
    (~cleaned_shooters['Fate of Shooter'].isnull()) & 
    ~(cleaned_shooters['Fate of Shooter'] == 'FALSE')
]

In [22]:
#Change 'Unknowns' to 'No' for history of mental illness (it's not a great assumption, but a necessary one)
cleaned_shooters['History of Mental Illness - General'].replace('Unknown', 'No', inplace=True)

#Change to True/False scheme
cleaned_shooters['History of Mental Illness - General'].replace('Yes', True, inplace=True)
cleaned_shooters['History of Mental Illness - General'].replace('No', False, inplace=True)

In [23]:
shooters['Place Type'].str.lower().unique()

array(['college/university/adult education', 'government facility',
       'primary school', 'secondary school',
       'retail/wholesale/services facility', 'entertainment venue',
       'restaurant/cafe?', 'company/factory/office',
       'retail/ wholesale/services facility\nand primary school',
       'public transportation', 'residential home/neighborhood',
       'military facility', 'street/highway', 'place of worship',
       'park/wildness', 'retail/ wholesale/services facility',
       'residential home/neighborhood \nand street/highway',
       'medical/care',
       'retail/wholesale/services facility\n/residential home/neighborhood',
       'residential home/neighborhood,\nretail/ wholesale/services facility',
       'restaurant/cafeé', 'restaurant/cafe', 'park/wilderness',
       'unknown', 'residential home'], dtype=object)

In [24]:
len(cleaned_shooters.loc[cleaned_shooters['Place Type'] == 'company/factory/office'])

0

In [25]:
#Cleaning/merging certain values so our data is simplified
cleaned_shooters['Place Type'] = shooters['Place Type'].str.lower()

In [26]:
#We'll iterate over our desired features and flatten them s.t.
#each unique value has a column, and if the incident has that value
#we'll just assign the matching column a 1 and the other ones a 0
marketbasket = cleaned_shooters[[
    'Number of Civilian Fatalities',
    'Number of Civilian Injured',
    'Number of Enforcement Fatalities',
    'Number of Enforcement Injured',
    'Military Experience',
    'Number of Shotguns',
    'Number of Rifles',
    'Number of Handguns',
    'High Civilian Casualties',
    'History of Mental Illness - General'
]]


#Get the logical inverses as well
marketbasket['No Civilian Fatalities'] = ~marketbasket['Number of Civilian Fatalities']
marketbasket['No Civilian Injured'] = ~marketbasket['Number of Civilian Injured']
marketbasket['No Enforcement Fatalities'] = ~marketbasket['Number of Enforcement Fatalities']
marketbasket['No Enforcement Injured'] = ~marketbasket['Number of Enforcement Injured']
marketbasket['No History of Mental Illness - General'] = ~marketbasket['History of Mental Illness - General']


##TODO: Need to finish generating the rest of the dataframe by expanding out unique values across columns
desired_features = [
    'Fate of Shooter',
    'Fate of Shooter at the scene',
    'Possible Motive - General'
]

for feature in desired_features:
    for val in cleaned_shooters[feature].unique():
        marketbasket[feature + ' ' + str(val)] = (cleaned_shooters[feature] == val)

marketbasket

C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexi

Number of Civilian Fatalities  Number of Civilian Injured  \
0                             True                        True   
1                             True                        True   
2                             True                        True   
3                             True                        True   
4                             True                        True   
5                             True                        True   
6                             True                        True   
7                             True                        True   
8                             True                        True   
9                             True                        True   
10                            True                        True   
11                            True                        True   
12                            True                        True   
13                            True                        True   
14                            True                        True   
15                            True                        True   
16                            True                        True   
17                            True                        True   
18                            True                        True   
19                            True                        True   
20                            True                        True   
21                            True                        True   
22                            True                        True   
23                            True                        True   
24                            True                        True   
25                            True                        True   
26                            True                        True   
27                           False                        True   
28                            True                        True   
29                            True                        True   
..                             ...                         ...   
305                           True                        True   
306                          False                        True   
307                           True                       False   
308                           True                        True   
309                          False                        True   
310                           True                       False   
311                           True                       False   
312                           True                        True   
313                           True                        True   
314                           True                        True   
315                           True                        True   
316                           True                        True   
317                           True                        True   
318                           True                        True   
319                           True                        True   
320                          False                        True   
321                           True                       False   
322                           True                       False   
323                           True                       False   
324                           True                        True   
325                           True                        True   
326                           True                        True   
327                           True                        True   
328                          False                        True   
329                           True                        True   
330                          False                        True   
331                           True                        True   
332                          False                        True   
333                           True           

In [27]:
#Apply apriori market basket algorithm
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
apriori_res = apriori(marketbasket, min_support=0.05, use_colnames=True)

In [28]:
#Create display copy of apriori results
apriori_res_display = apriori_res.copy(deep=True)
apriori_res_display['itemsets'] = apriori_res_display['itemsets'].apply(lambda x: ''.join(e + ', ' for e in x))

In [29]:
pd.options.display.max_colwidth = 300
apriori_res_display.sort_values('support')

support  \
958   0.051205   
941   0.051205   
908   0.051205   
1849  0.051205   
1860  0.051205   
878   0.051205   
1870  0.051205   
1871  0.051205   
858   0.051205   
852   0.051205   
185   0.051205   
1877  0.051205   
1880  0.051205   
1885  0.051205   
193   0.051205   
194   0.051205   
811   0.051205   
810   0.051205   
807   0.051205   
798   0.051205   
797   0.051205   
794   0.051205   
784   0.051205   
780   0.051205   
775   0.051205   
770   0.051205   
948   0.051205   
769   0.051205   
1810  0.051205   
968   0.051205   
...        ...   
918   0.524096   
1106  0.527108   
398   0.533133   
291   0.533133   
100   0.542169   
279   0.542169   
99    0.548193   
386   0.548193   
36    0.557229   
57    0.560241   
6     0.575301   
733   0.599398   
204   0.614458   
203   0.626506   
26    0.650602   
606   0.659639   
157   0.668675   
144   0.689759   
13    0.710843   
385   0.743976   
56    0.759036   
278   0.771084   
55    0.780120   
35    0.789157   
34    0.804217   
1     0.807229   
0     0.834337   
143   0.915663   
12    0.936747   
11    0.960843   

                                                                                                                                                                  itemsets  
958                            Number of Civilian Fatalities, No Enforcement Injured, No History of Mental Illness - General, Possible Motive - General Domestic dispute,   
941                                       Number of Civilian Fatalities, No Enforcement Fatalities, Fate of Shooter Deceased, Possible Motive - General Domestic dispute,   
908                                                     Number of Civilian Fatalities, No Civilian Injured, No Enforcement Injured, Fate of Shooter at the scene Custody,   
1849             Number of Civilian Injured, No Enforcement Fatalities, No Enforcement Injured, Fate of Shooter at the scene Deceased, Possible Motive - General Unknown,   
1860  Number of Civilian Injured, No Enforcement Fatalities, Fate of Shooter Deceased, Fate of Shooter at the scene Deceased, Possible Motive - General Multiple motives,   
878                                              Number of Civilian Fatalities, High Civilian Casualties, No Enforcement Injured, No History of Mental Illness - General,   
1870              Number of Civilian Injured, No Enforcement Injured, Fate of Shooter Deceased, Fate of Shooter at the scene Deceased, Possible Motive - General Unknown,   
1871         Number of Civilian Injured, No Enforcement Injured, Fate of Shooter Custody, Fate of Shooter at the scene Custody, Possible Motive - General Social dispute,   
858                                                  Number of Civilian Fatalities, Number of Handguns, No Enforcement Injured, Possible Motive - General Mental illness,   
852                                             Number of Civilian Fatalities, Number of Handguns, No Enforcement Fatalities, Possible Motive - General Domestic dispute,   
185                                                                                                    Fate of Shooter Custody, Possible Motive - General Mental illness,   
1877                                                   Number of Rifles, Number of Handguns, No Enforcement Fatalities, No Enforcement Injured, Fate of Shooter Deceased,   
1880                               Number of Rifles, No Enforcement Fatalities, No Enforcement Injured, No History of Mental Illness - General, Fate of Shooter Deceased,   
1885                   Number of Rifles, No Enforcement Injured, No History of Mental Illness - General, Fate of Shooter Deceased, Fate of Shooter at the scene Deceased,   
193                                                                                       Fate of Shooter at the scene Custody, Possible Motive - General Mental illness,   
194                                                                                       Fate of Shooter at the scene Cus

In [30]:
#Display full dataset for apriori results, limited to itemsets of <=3 elements
max_rows_orig = pd.options.display.max_rows
pd.options.display.max_rows = 1000
apriori_res.loc[[len(itemset) <= 3 for itemset in apriori_res['itemsets']]]

support  \
0    0.834337   
1    0.807229   
2    0.063253   
3    0.111446   
4    0.138554   
5    0.234940   
6    0.575301   
7    0.153614   
8    0.289157   
9    0.165663   
10   0.192771   
11   0.960843   
12   0.936747   
13   0.710843   
14   0.433735   
15   0.445783   
16   0.102410   
17   0.424699   
18   0.367470   
19   0.195783   
20   0.099398   
21   0.084337   
22   0.292169   
23   0.099398   
24   0.066265   
25   0.081325   
26   0.650602   
27   0.108434   
28   0.129518   
29   0.213855   
30   0.493976   
31   0.150602   
32   0.277108   
33   0.183735   
34   0.804217   
35   0.789157   
36   0.557229   
37   0.400602   
38   0.361446   
39   0.063253   
40   0.394578   
41   0.304217   
42   0.123494   
43   0.093373   
44   0.246988   
45   0.096386   
46   0.063253   
47   0.075301   
48   0.081325   
49   0.120482   
50   0.201807   
51   0.463855   
52   0.153614   
53   0.246988   
54   0.156627   
55   0.780120   
56   0.759036   
57   0.560241   
58   0.310241   
59   0.379518   
60   0.099398   
61   0.301205   
62   0.310241   
63   0.183735   
64   0.084337   
65   0.081325   
66   0.216867   
67   0.084337   
68   0.054217   
69   0.075301   
70   0.099398   
71   0.093373   
72   0.063253   
73   0.075301   
74   0.072289   
75   0.057229   
76   0.057229   
77   0.069277   
78   0.126506   
79   0.126506   
80   0.069277   
81   0.081325   
82   0.057229   
83   0.081325   
84   0.054217   
85   0.117470   
86   0.075301   
87   0.099398   
88   0.210843   
89   0.198795   
90   0.135542   
91   0.117470   
92   0.111446   
93   0.111446   
94   0.096386   
95   0.123494   
96   0.201807   
97   0.081325   
98   0.111446   
99   0.548193   
100  0.542169   
101  0.373494   
102  0.289157   
103  0.271084   
104  0.283133   
105  0.228916   
106  0.057229   
107  0.063253   
108  0.054217   
109  0.114458   
110  0.072289   
111  0.051205   
112  0.051205   
113  0.096386   
114  0.141566   
115  0.144578   
116  0.057229   
117  0.102410   
118  0.051205   
119  0.099398   
120  0.051205   
121  0.271084   
122  0.268072   
123  0.156627   
124  0.132530   
125  0.153614   
126  0.123494   
127  0.093373   
128  0.057229   
129  0.156627   
130  0.147590   
131  0.153614   
132  0.084337   
133  0.063253   
134  0.072289   
135  0.180723   
136  0.177711   
137  0.150602   
138  0.123494   
139  0.066265   
140  0.123494   
141  0.057229   
142  0.075301   
143  0.915663   
144  0.689759   
145  0.406627   
146  0.433735   
147  0.102410   
148  0.400602   
149  0.355422   
150  0.192771   
151  0.087349   
152  0.081325   
153  0.292169   
154  0.099398   
155  0.063253   
156  0.075301   
157  0.668675   
158  0.391566   
159  0.424699   
160  0.102410   
161  0.385542   
162  0.346386   
163  0.192771   
164  0.090361   
165  0.084337   
166  0.280120   
167  0.093373   
168  0.057229   
169  0.066265   
170  0.277108   
171  0.313253   
172  0.102410   
173  0.271084   
174  0.243976   
175  0.183735   
176  0.078313   
177  0.253012   
178  0.066265   
179  0.424699   
180  0.117470   
181  0.057229   
182  0.057229   
183  0.367470   
184  0.078313   
185  0.051205   
186  0.066265   
187  0.093373   
188  0.102410   
189  0.075301   
190  0.114458   
191  0.057229   
192  0.057229   
193  0.051205   
194  0.051205   
195  0.057229   
196  0.114458   
197  0.078313   
198  0.111446   
199  0.186747   
200  0.385542   
201  0.150602   
202  0.237952   
203  0.626506   
204  0.614458   
205  0.412651   
206  0.286145   
207  0.295181   
208  0.060241   
209  0.280120   
210  0.246988   
211  0.111446   
212  0.081325   
213  0.171687   
214  0.081325   
215  0.054217   
216  0.072289   
217  0.096386   
218  0.090361   
219  0.060241   
220  0.072289   
221  0.069277   
222  0.057229   
223  0.054217   
224  0.066265   
225  0.117470   
226  0.120482   
227  0.063253   
228  0.075301   
229  0.054217   
230  0.075301   
231  0.051205   
232  0.111446   
233  0.075301   
234  0.0993

In [31]:
#Reset number of rows that can be shown
pd.options.display.max_rows = max_rows_orig

Great, all of our data is wrapped up in the marketbasket and we have apriori support values for all potential combinations of different variables down to the case where 5% of incidents exhibited a given combination. 

I'm going to start by drilling down into data involving members of the military. It's an interesting twist that someone who would fight to defend the country would turn their gun on the general population. We'll see what shakes out here.

In [32]:
#How many incidents had a shooter w/ military experience?
len(marketbasket.loc[marketbasket['Military Experience'] == True])

37

In [33]:
#Select market basket results with 'Military Experience'
military_apriori = apriori_res.loc[['Military Experience' in elem for elem in apriori_res['itemsets']]]
military_apriori

support  \
3     0.111446   
27    0.108434   
48    0.081325   
69    0.075301   
70    0.099398   
71    0.093373   
72    0.063253   
73    0.075301   
74    0.072289   
197   0.078313   
216   0.072289   
217   0.096386   
218   0.090361   
219   0.060241   
220   0.072289   
221   0.069277   
326   0.057229   
327   0.072289   
328   0.069277   
329   0.060241   
330   0.057229   
431   0.066265   
432   0.060241   
433   0.057229   
434   0.054217   
435   0.090361   
436   0.057229   
437   0.066265   
438   0.066265   
439   0.054217   
...        ...   
777   0.054217   
778   0.063253   
779   0.063253   
780   0.051205   
781   0.057229   
782   0.057229   
783   0.069277   
983   0.066265   
984   0.051205   
985   0.051205   
986   0.057229   
1169  0.057229   
1170  0.054217   
1171  0.054217   
1172  0.057229   
1173  0.057229   
1174  0.066265   
1175  0.060241   
1397  0.063253   
1398  0.054217   
1538  0.054217   
1539  0.051205   
1540  0.051205   
1541  0.054217   
1542  0.054217   
1543  0.063253   
1544  0.057229   
1722  0.051205   
1874  0.057229   
2085  0.054217   

                                                                                                                                                                      itemsets  
3                                                                                                                                                        [Military Experience]  
27                                                                                                                        [Number of Civilian Fatalities, Military Experience]  
48                                                                                                                           [Number of Civilian Injured, Military Experience]  
69                                                                                                                                   [Military Experience, Number of Handguns]  
70                                                                                                                            [Military Experience, No Enforcement Fatalities]  
71                                                                                                                               [Military Experience, No Enforcement Injured]  
72                                                                                                               [Military Experience, No History of Mental Illness - General]  
73                                                                                                                             [Military Experience, Fate of Shooter Deceased]  
74                                                                                                                [Military Experience, Fate of Shooter at the scene Deceased]  
197                                                                                           [Number of Civilian Fatalities, Number of Civilian Injured, Military Experience]  
216                                                                                                   [Number of Civilian Fatalities, Military Experience, Number of Handguns]  
217                                                                                            [Number of Civilian Fatalities, Military Experience, No Enforcement Fatalities]  
218                                                                                               [Number of Civilian Fatalities, Military Experience, No Enforcement Injured]  
219                                                                               [Number of Civilian Fatalities, Military Experience, No History of Mental Illness - General]  
220                                                                                             [Number of Civilian Fatalities, Military Experience, Fate of Shooter Deceased]  
221                                                       

In [34]:
#Number of cases where the shooter had military experience and died on the scene
len(marketbasket.loc[(marketbasket['Military Experience'] == True) & (marketbasket['Fate of Shooter at the scene Deceased'] == True)])

24

In [42]:
#How often do shooters w/o military experience die on the scene?
deceased_no_military_experience = marketbasket.loc[
    (marketbasket['Fate of Shooter at the scene Deceased'] == True) &
    (marketbasket['Military Experience'] == False)
]
no_military_experience = marketbasket.loc[marketbasket['Military Experience'] == False]
float(len(deceased_no_military_experience)) / float(len(no_military_experience))

0.39661016949152544

In [36]:
%load_ext rpy2.ipython

The rpy2.ipython extension is already loaded. To reload it, use:
  %reload_ext rpy2.ipython


In [43]:
%%R
#Test for statistical significance
cm1 <- matrix(c(24, 117, 13, 178),
                nrow = 2,
                dimnames = list(c("Military", "Not Military"), c("Deceased", "Not Deceased"))) 
addmargins(as.matrix(cm1), c(1,2))
fisher.test(cm1)

From the command line:


Fisher's Exact Test for Count Data<br/>
data:  cm1<br/>
**p-value = 0.004469**<br/>
alternative hypothesis: true odds ratio is not equal to 1<br/>
95 percent confidence interval:<br/>
 1.309101 6.243371<br/>
sample estimates:<br/>
odds ratio<br/>
  2.799742<br/>


**In other words, our observed difference is very statistically significant.**

Here's a breakdown of the results from my military experience analysis:

- 37 of the 335 incidents in question involved a shooter with military experience (~11%).
    - According to an analysis by FiveThirtyEight, it's estimated that 7.3% of all living Americans served in the armed forces.
    - I'm going to chalk up that gap in percentage of shooters relative to the overall population to small sample size. There's no way to know exactly how many mass shootings have occurred since 1960, but if we had all of them then I'd guess that this gap would shrink.
    
- **In 64.8% of incidents where the shooter had military experience, they died on the scene**
    - That's a huge uptick from the general case: 43.4% of all shooters die on the scene
    - 39.7% of shooters with no military experience die on the scene
    - According to the Fisher test, our p-value for that difference is .004469; *very* statistically significant.
    - Military experience seems to be a large contributing factor in why someone would die taking a final stand.
    - In 79% of these cases, no officers were injured or killed. How did these guys provoke the cops to the point where they were killed on the spot without actually injuring or killing any of them?
    
- 81% of cases w/ military experience result in officer injury or death
    - General case: 91.6% of cases result in no enforcement injuries or deaths
    - But again, a 10% shift in this subset is a matter of 3-4 cases, probably not enough to go off of
    - In any event, it's surprising that a mass shooter with military experience wouldn't target law enforcement if their actions are in any way related to an institutional resentment
        - But then again I should probably drill down into that; what were the motives for the military guys?

In [ ]:
num_officers_not_targeted = len(shooters.loc[(shooters['Number of Enforcement Fatalities'] == 0) & (shooters['Number of Enforcement Injured'] == 0)])
print("Officers were injured or killed ", float(len(shooters) - num_officers_not_targeted) / float (len(shooters)), " of the time.")

Interesting. Nearly every example of a shooting where they had military experience resulted in no officer casualties. That being said, it looks roughly proportional to the number of shooters without military experience that did/didn't target law enforcement. In other words, military experience doesn't appear to make a significant difference in who they choose to target.

In [ ]:
from builtins import any as b_any

#Get motive results, remove ones where we don't know their motive or there are multiple ill-defined motives
motive_apriori = apriori_res.loc[[b_any('Possible Motive - General' in e for e in elem) for elem in apriori_res['itemsets']]]
# motive_apriori = motive_apriori[['Possible Motive - General Unknown' not in elem for elem in motive_apriori['itemsets']]]
# motive_apriori = motive_apriori[['Possible Motive - General Multiple motives' not in elem for elem in motive_apriori['itemsets']]]
motive_apriori
# motive_apriori.sort_values('support')

The market basket analysis gives us the following insights:
- Law enforcement is injured in just ~6.3% of mass shootings; officers die in 3.9% of incidents.
- Officers were neither injured nor killed in 307 out of 335 incidents (91.6%).
    - Is that discrepancy because law enforcement is adequately trained for these kinds of situations?
    - Or is it because the shooters aren't actively targeting law enforcement?
    - I'm guessing the latter, because the analysis points out that a civilian dies in 83.6% of shootings; if they were targeting law enforcement too, they'd probably have a higher 'success' rate
- 11% of the shooters had military experience
- There's a papertrail for mental illness in only 28.9% of cases

In [ ]:
#Group the dataset by year
shooters['Date'] = pd.to_datetime(shooters['Date'])
dategroups = shooters.groupby(shooters.Date.dt.year)


#Iterate over groups, count each gun type
for name, grouped in dategroups:
    gun_counts = grouped['Type of Gun - General'].count().transpose()
    print(name)
    print(gun_counts)
    


In [ ]:
from altair import *
chart = Chart(data).mark_area(
    stacked='center',
).encode(
    color=Color('series:N',
        scale=Scale(
            range='category20b',
        ),
    ),
    x=X('date:T',
        axis=Axis(
            axisWidth=0.0,
            format='%Y',
            labelAngle=0.0,
            tickSize=0.0,
        ),
        scale=Scale(
            nice='month',
        ),
        timeUnit='yearmonth',
    ),
    y=Y('sum(count):Q',
        axis=False,
    ),
).configure_cell(
    height=200.0,
    width=300.0,
)

Alright, here's an idea: let's look at the density of different types of shootings across the country based on motive first, then other variables.

In [ ]:
#Let's start by looking at how many examples there are for each motive
shooters['Possible Motive - General'] = shooters['Possible Motive - General'].str.lower()
motive_groups = shooters.groupby('Possible Motive - General')
for name, group in motive_groups:
    print(name, ": ", len(group))

I suppose it's a blessing that there are too few examples of any one kind of shooting to use. In any event, it means that standard density estimation techniques are out the window, except when considering all examples.

In [ ]:
from mpl_toolkits.basemap import Basemap

COLORS=['#1DACE8','#1C366B','#F24D29','#E5C4A1','#C4CFD0']
SELECTED_GROUPS = [
    'social dispute',
    'terminated/denied/reprimanded',
    'domestic dispute',
    'mental illness',
    'political/religious ideals'
]

fig, ax = plt.subplots()
shooter_map = Basemap(llcrnrlon=-145.5,llcrnrlat=1.,urcrnrlon=-2.566,urcrnrlat=46.352,\
            rsphere=(6378137.00,6356752.3142),
            resolution='l',area_thresh=1000.,projection='merc',
            lat_1=50.,lon_0=-107.,ax=ax)
shooter_map.drawcoastlines()
shooter_map.drawcountries()
shooter_map.fillcontinents(color = 'black')
shooter_map.drawmapboundary()


##TODO: fix the coordinates and map projection

#Plot the shooting motivation groups we've indicated above
color_counter = 0
for name, group in motive_groups:
    if name in SELECTED_GROUPS:
        shooter_map.plot(
            group['Latitude'], 
            group['Longitude'],
#             color=COLORS[color_counter],
            'bo',
            markersize=20
        )
        color_counter += 1

plt.show()